### PR TITLE
Add feature gates for `for` and `?` in consts

### PR DESCRIPTION
**Pull request by jonas-schievink**
These operations seems *relatively* straightforward to support, and only seem to be blocked on `impl const Trait`.

I have included a working test for `const_try`, but `const_for` is currently unusable without reimplementing *every single* defaulted `Iterator` method, so I didn't do that.

(both features still need tracking issues before this is merged)

### CODE REVIEW

**Comment by oli-obk** [comment]
> but `const_for` is currently unusable without reimplementing _every single_ defaulted `Iterator` method, so I didn't do that.

maybe https://github.com/rust-lang/rust/pull/86857 helps here?

**Comment by jonas-schievink** [comment]
In theory, yeah, this works:

```rs
#![feature(const_refs_to_cell)]
#![feature(const_mut_refs)]
#![feature(const_fn_trait_bound)]
#![feature(const_trait_impl)]

trait Iter: Copy + Sized {
    type Item: Copy;
    
    fn next(&mut self) -> Option<Self::Item>;
    
    #[default_method_body_is_const]
    fn count(mut self) -> usize {
        let mut count = 0;
        while self.next().is_some() {
            count += 1;
        }
        count
    }
}

#[derive(Copy, Clone)]
struct ConstIter {
    next: Option<u8>,
}

impl const Iter for ConstIter {
    type Item = u8;
    
    fn next(&mut self) -> Option<u8> {
        match self.next {
            Some(val) => {
                self.next = val.checked_add(1);
                Some(val)
            }
            None => None,
        }
    }
}

const _: () = {
    let c = ConstIter { next: Some(0) }.count();
    if c != 256 {
        [()][4];
    }
};
```

...but it requires a lot of feature gates and `Copy` bounds (since everything is generic) that wouldn't work on `Iterator`. Seems like there's still a bunch of work to do until iterators work in consts.

**Comment by oli-obk** [comment]
Then I would say we don't add a const_for feature gate, we can start figuring out const iterators without it.

**Comment by jonas-schievink** [comment]
But we'll need one eventually, right? Why not add it now?

**Comment by oli-obk** [comment]
Yea. Just right now it doesn't really get us anything actionable. I guess as long as we have a canary test showing the behaviour of the feature gate it should be ok to have an unusable feature gate. Just thought it unusual.

**Comment by usbalbin** [comment]
#86853 is a PR which might have some overlap with the try-operator part of this. My goal there was to try and constify the most trivial conversion impls to enable basic usage with `Option` and `Result` (however that turned out to be blocked on some more work on `const_trait_impl`)

**Comment by usbalbin** [comment]
Btw is #87576 a duplicate of #74935? :)

**Comment by jonas-schievink** [comment]
Yes, these tracking issues are pretty hard to discover

File: compiler/rustc_feature/src/active.rs [hunk 1]
```diff
@@ -677,6 +677,12 @@ declare_features! (
     /// Infer generic args for both consts and types.
     (active, generic_arg_infer, "1.55.0", Some(85077), None),
 
+    /// Allows `for _ in _` loops in const contexts.
+    (active, const_for, "1.55.0", Some(87575), None),
```
Comment (oli-obk) [inline compiler/rustc_feature/src/active.rs:681]: we are now in 1.56 territory

File: src/test/ui/consts/const-for.stderr [hunk 2]
```diff
@@ -0,0 +1,15 @@
+error[E0015]: calls in constants are limited to constant functions, tuple structs and tuple variants
+  --> $DIR/const-for.rs:5:14
+   |
+LL |     for _ in 0..5 {}
+   |              ^^^^
```
Comment (oli-obk) [inline src/test/ui/consts/const-for.stderr:5]: I guess we should start printing the function's name in the diagnostic. Very uninformative this error XD

**Pull request merged**